logic/src/commonMain/kotlin/com/wire/kalium/logic/data/asset/AssetRepository.kt

```diff
@@ -164,30 +183,39 @@ internal class AssetDataSource(
         decodedDataPath: Path,
         extension: String?
     ): Either<CoreFailure, UploadedAssetId> =
-        assetMapper.toMetadataApiModel(uploadAssetData, kaliumFileSystem).let { metaData ->
-            wrapApiRequest {
-                val dataSource = kaliumFileSystem.source(uploadAssetData.tempEncryptedDataPath)
-
-                // we should also consider for avatar images, the compression for preview vs complete picture
-                assetApi.uploadAsset(metaData, { dataSource }, uploadAssetData.dataSize).also { dataSource.close() }
+        assetMapper.toMetadataApiModel(uploadAssetData, kaliumFileSystem)
+            .let { metaData ->
+                wrapApiRequest {
+                    val dataSource = kaliumFileSystem.source(uploadAssetData.tempEncryptedDataPath)
+
+                    // we should also consider for avatar images, the compression for preview vs complete picture
+                    assetApi.uploadAsset(metaData, { dataSource }, uploadAssetData.dataSize)
+                        .also { dataSource.close() }
+                }
+            }
+            .flatMap { assetResponse ->
+                // After successful upload, we persist the asset to a persistent path
+                persistAsset(assetResponse.key, assetResponse.domain, decodedDataPath, uploadAssetData.dataSize, extension)
+                    .also { kaliumFileSystem.delete(uploadAssetData.tempEncryptedDataPath) }
```

Comment:
Should we delete path when persistAsset() fails?

Reply:
For uploading private assets, this is a temporary file located in cache directory so it will be removed anyway sooner or later but when we try to do it again, we will execute `uploadAndPersistPrivateAsset` which will store it in a cache again before sending so it's safer to clear any files before retrying. For public assets we probably won't have the retry mechanism so after it fails, we will need to do that all over again which also means create another temporary file so it's probably ok to delete it also when it fails.